### ISSUE TITLE
Suppress `:unused-binding` linter in first argument of class method arglists

### ISSUE THREAD

**Issue by yuhan0** [clj-kondo/clj-kondo#2378]
To upvote this issue, give it a thumbs up. See [this list](https://github.com/clj-kondo/clj-kondo/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc) for the most upvoted issues.

- [x] I have read the [Clojure etiquette](https://clojure.org/community/etiquette) and will respect it when communicating on this platform.

**Is your feature request related to a problem? Please describe.**

When defining protocol or interface implementations in a `defrecord` / `deftype` form, the compulsory first parameter to the method (usually called `obj` / `this` / `self`) may appear to be unused in the body, causing the `:unused-binding` linter to produce a warning:

```clj
(defrecord Foo [bar]
  Bazzable
  (baz [this] bar)) ;; <- Warning: unused binding this
```

The binding is in fact being used, because `bar` acts as an implicit field reference on the class instance passed to `this` parameter, i.e. can be treated as `(baz [this] (.-bar this))`

**Describe the solution you'd like**

The `:unused-binding` linter should ignore the first argument to method implementation arglists in `defrecord` / `deftype` forms, and possibly also in `extend` / `extend-protocol` / `extend-type` / `reify`.

**Describe alternatives you've considered**

### Rename binding with leading underscore

Using `this` as an arg name is idiomatic, having to rename it to `_this` / `_` to satisfy the linter seems like a poor workaround

### Analyze defrecord / deftype fields as explicit (.-field this) expressions
This lets the linter catch situations where the class instance passed to the method actually goes unused:

```clj
(defrecord Foo [bar]
  Bazzable
  (baz [this] bar)    ;; <no warning>
  (qux? [this] true)) ;; Warning: unused binding this
```
This is arguably more accurate, but I would personally prefer for the linter to be always suppressed on the first arg. This allowing for consistent naming across methods, rather than alternating `this` and `_this` depending on whether the method body happens to involve one of the fields.

Sometimes the compulsory presence of the argument can be treated as a sort of boilerplate, and having a consistent naming style is helpful for clarity, even if it goes unused.

This also applies in the case of `reify` and `extend-*` , where there are no field bindings in scope as with `defrecord`, but I'd still prefer for the linter not to be triggered. E.g. the example code given in `reify`'s docstring:

```clojure
(let [f "foo"] 
  (reify clojure.lang.Seqable
    (seq [this] (seq f))))
```

### Suppress the unused-binding linter altogether 
Misses out on actual unused bindings (too coarse)
```clojure
(deftype Foo [bar]
  Object
  #_{:clj-kondo/ignore [:unused-binding]}
  (equals [this that] true)) ;; <- expected: unused binding `that`
```

**Additional context**
cmdline invocation:
```bash
$ echo '(deftype Foo [bar] Object (toString [this] bar))' | clj-kondo --lint -
<stdin>:1:38: warning: unused binding this
```

Hmm.. what if destructuring is involved? Should the linter be triggered in the following:
```clj
(defrecord Foo [bar]
  Quxxable
  (qux [{:keys [what now]}] what))
```
(surely nobody does this anyway)

**Comment by borkdude** [comment]
Proposal: We could either just not warn on the unused this binding by default and make an option to restore the old behavior, or vice versa.

**Comment by borkdude** [comment]
Someone posted on Slack that you can do this with:

``` clojure
{:linters {:unused-binding {:exclude-patterns ["^this"]}}}
```

Also: just writing `(qux [_ a b c])` is perfectly idiomatic if you ask me. So perhaps we already have enough solutions to not justify extra complexity

Re: destructuring, since the destructuring is completely unnecessary when you don't use those bindings, I'd say a warning is justified.

**Comment by yuhan0** [comment]
Ahh yes, it was even mentioned specifically in the config docs 🤦

https://github.com/clj-kondo/clj-kondo/blob/30ab602bbfb0381b164c73f67aaef26727a5bb6d/doc/linters.md?plain=1#L1653

Although it would still be a nice-to-have feature (I tend to give the arg recognizable names depending on the protocol, instead of `this`), I agree it's low priority given the `:exclude-binding` solution. Thanks!
